pandas/core/groupby/grouper.py

```diff
@@ -573,7 +571,7 @@ def _get_grouper(
             all_in_columns_index = all(
                 g in obj.columns or g in obj.index.names for g in keys
             )
-        elif isinstance(obj, Series):
+        else:  # Series
```

Comment:
Hmm is there a reason to remove this isinstance check? Seems to loosen the type checker

Reply:
I had the reverse problem where no else clause made the type unknown. An alternative would be to do:

```python
elif isinstance(obj, Series):
    ....
else:
  raise TypeError(...)
```
?

Reply:
the next line accesses obj.index, so implicitly pins it down.  the comment above is fine or an assertion if that helps mypy out

Reply:
Not sure I read you comment clearly. The chosen solution os ok for you ( if not preferred:-))?

Reply:
I think assert isinstance(obj, Series) within the branch should work then. Sounds like it doesn't make much of a difference with what is hitting this path now, but the original intent is definitely only for Series objects to make it down this path

Reply:
not checked, but I assume that mypy can't perform type narrowing between a DataFrame and Series since we use NDFrame as the bound for FrameOrSeries.

so we know that if obj is not a DataFrame then it must be a Series, but mypy can't know that.

if it's a public method, and we don't have control of the calling types then https://github.com/pandas-dev/pandas/pull/29458#discussion_r343736687 makes sense